Pro Dashboard: Add a reusable contact list component to unify the implementation of email and SMS in notification settings.

jkguidaven opened the pull request:
This is part of my ongoing refactoring of the Email and Phone number editor to have shared codes.

This PR adds a reusable Contact List component to unify the implementation of the Email and SMS contact list in the Notification settings modal.

<img width="514" alt="Screen Shot 2023-06-23 at 4 16 55 PM" src="https://github.com/Automattic/wp-calypso/assets/56598660/339f67ad-0b4d-40e9-9452-6847294533c2">


Related to 1204774821045518-as-1204802449809227

## Proposed Changes

## Testing Instructions

Since these changes are made specifically for agencies, you must set yourself(partner) as an agency - 2c49b-pb. Make sure to switch it back to the previous type.

**Instructions**

* Run `git checkout add/downtime-monitoring-contact-list-common-component` and `yarn start-jetpack-cloud`
* Open http://jetpack.cloud.localhost:3000/, and you'll be redirected to the /dashboard.
* Enable monitor if not enabled already.
* Click on the clock icon next to the monitor toggle.
* Confirm that the email and SMS contact list work as expected.

## Pre-merge Checklist


- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

jkguidaven committed:
Add contact list component.

jkguidaven committed:
Update existing components to use the new contact list.

jkguidaven committed:
Fix incorrect import.

jkguidaven committed:
Fix contact list button class name.

jkguidaven committed:
Rename contact-list-item.tsx to item.tsx

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-list/index.tsx:
```diff
@@ -0,0 +1,71 @@
+				<div className="margin-top-16">
+					<AlertBanner type="warning">
+						{ type === 'email'
+							? translate( 'You need at least one email address' )
```
We won't need this for `email`.

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-list/index.tsx:
```diff
@@ -0,0 +1,71 @@
+
+				<Button compact className="contact-list__button" onClick={ onAddContact }>
+					<Icon size={ 18 } icon={ plus } />
+					{ type === 'email' ? translate( 'Add email address' ) : translate( 'Add phone number' ) }
```
Can we also check if the `type` is `phone` here? Let's not have a default one considering we will have other modes of notifications.

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-list/utils.ts:
```diff
@@ -0,0 +1,37 @@
+	type: 'email' | 'phone',
+	action: AllowedMonitorContactActions
+) => {
+	const EVENT_NAMES =
```
Can we also check if the `type` is `phone` here? Let's not have a default one considering we will have other modes of notifications. 

Something like

```
EVENT_NAMES = {
  email: {
   }
   phone: {
   }
}
```

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-list/index.tsx:
```diff
@@ -0,0 +1,71 @@
+			) }
+
+			<div className="contact-list">
+				{ items.map( ( item: StateMonitorSettingsEmail | StateMonitorSettingsSMS ) => (
```
I don't think we have to specify the type here.

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/contact-list/utils.ts:
```diff
@@ -0,0 +1,37 @@
+	type: 'email' | 'phone',
+	item: StateMonitorSettingsEmail | StateMonitorSettingsSMS
+) => {
+	if ( type === 'email' ) {
```
Can we also check if the `type` is phone here? Let's not have a default one considering we will have other modes of notifications.

jkguidaven committed:
Add 'AllowedMonitorContactTypes' enum.

yashwin reviewed (commented):
Thanks for working on creating the common components. 

I found an issue here

I don't see any options when I click the `...` icon. Is that expected? 

<img width="264" alt="Screenshot 2023-06-23 at 3 01 51 PM" src="https://github.com/Automattic/wp-calypso/assets/10586875/bda5599c-a4cd-48eb-aaad-19f3669fb33e">


I have also left some comments.

jkguidaven committed:
Fix PR comments.

jkguidaven commented:
> Thanks for working on creating the common components.
> 
> I found an issue here
> 
> I don't see any options when I click the `...` icon. Is that expected?
> 
> <img alt="Screenshot 2023-06-23 at 3 01 51 PM" width="264" src="https://user-images.githubusercontent.com/10586875/248231416-bda5599c-a4cd-48eb-aaad-19f3669fb33e.png">
> I have also left some comments.

@yashwin Thanks for reviewing the PR. 🙏  I addressed all the PR comments and hope I didn't miss anything. 

Thanks for catching the popover menu bug. For some reason, it was working on my local machine until I stopped running the Calypso and doing `yarn start-jetpack-cloud` again.

jkguidaven committed:
Fix warning message showing up

jkguidaven commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/types.ts:
```diff
@@ -328,8 +328,10 @@ export type MonitorSettingsContact = Partial< MonitorSettingsEmail > &
 
 export type AllowedMonitorContactActions = 'add' | 'verify' | 'edit' | 'remove';
 
+export type AllowedMonitorContactTypes = 'email' | 'sms';
```
I also added allowable contact types and set them to 'email' and 'sms' to match our endpoint. This makes sure we are using the correct strings.

yashwin reviewed (approved):
LGTM 🚀 Great job on this, @jkguidaven.
